wasm-runtime/polkadot/src/support/storable.rs

```diff
@@ -17,29 +17,48 @@
 //! Stuff to do with the runtime's storage.
 
 use slicable::Slicable;
-use endiansensitive::EndianSensitive;
 use keyedvec::KeyedVec;
 use runtime_support::prelude::*;
 use runtime_support::{self, twox_128};
 
 /// Trait for a value which may be stored in the storage DB.
 pub trait Storable {
 	/// Lookup the value in storage and deserialise, giving a default value if not found.
-	fn lookup_default(key: &[u8]) -> Self where Self: Sized + Default { Self::lookup(key).unwrap_or_else(Default::default) }
+	fn lookup_default(key: &[u8]) -> Self where Self: Sized + Default {
+		Self::lookup(key).unwrap_or_else(Default::default)
+	}
+
 	/// Lookup `Some` value in storage and deserialise; `None` if it's not there.
-	fn lookup(_key: &[u8]) -> Option<Self> where Self: Sized { unimplemented!() }
+	fn lookup(_key: &[u8]) -> Option<Self> where Self: Sized {
+		unimplemented!()
+	}
+
+	/// Retrives and returns the serialised value of a key from storage, removing it immediately.
+	fn take(key: &[u8]) -> Option<Self> where Self: Sized {
+		if let Some(value) = Self::lookup(key) {
+			kill(key);
+			Some(value)
+		} else {
+			None
+		}
+	}
+
 	/// Place the value in storage under `key`.
 	fn store(&self, key: &[u8]);
```

Comment:
having this as an implicit function on the objects is a little weird. You end up with code like `value.store("some_data")` which reads a more like the value is having data added to it.

invocations like `storage::store("some_key", &value)` and `storage::load("some_key")` better indicate that implicit globals are being accessed.

Reply:
although really this means that every object that implements this trait has to manually perform storage, when really the only common functionality needed is serialization/deserialization.

Reply:
i'm happy with:

```
let count: usize = database::lookup_default(key);
database::put(key, count + 1);
```

but i definitely don't think this is good:

```
let value_bytes = database::lookup_default(key);
let count = <usize>::deserialise(value_bytes);
let value_bytes = database::serialise(count + 1);
database::store(key, &value_bytes);
```